File: medium/1871-jump-game-vii/solution.py
class Solution:
    def canReach(self, s: str, minJump: int, maxJump: int) -> bool:
        """
        1871. Jump Game VII
        Time: O(n)
        Space: O(n)
        """
        n = len(s)
        reachable = [False] * n
        reachable[0] = True
        reachable_count = 0

        for index in range(1, n):
            add_index = index - minJump
            if add_index >= 0 and reachable[add_index]:
                reachable_count += 1

            remove_index = index - maxJump - 1
            if remove_index >= 0 and reachable[remove_index]:
                reachable_count -= 1

            reachable[index] = s[index] == "0" and reachable_count > 0

        return reachable[-1]

File: medium/1871-jump-game-vii/solution.java
/**
 * 1871. Jump Game VII
 * Time: O(n)
 * Space: O(n)
 */
class Solution {
    public boolean canReach(String s, int minJump, int maxJump) {
        int n = s.length();
        boolean[] reachable = new boolean[n];
        reachable[0] = true;
        int reachableCount = 0;

        for (int index = 1; index < n; index++) {
            int addIndex = index - minJump;
            if (addIndex >= 0 && reachable[addIndex]) {
                reachableCount++;
            }

            int removeIndex = index - maxJump - 1;
            if (removeIndex >= 0 && reachable[removeIndex]) {
                reachableCount--;
            }

            reachable[index] = s.charAt(index) == '0' && reachableCount > 0;
        }

        return reachable[n - 1];
    }
}
